File: app/api/evacuation.py
"""
Emergency Evacuation API

Provides endpoints for emergency response and evacuation management:
- Real-time list of all on-site visitors
- Evacuation mode trigger
- Muster point check-in
- Headcount and status tracking
"""
from flask import Blueprint, request, jsonify
from bson import ObjectId
from datetime import datetime, timezone
from app.db import visit_collection, visitor_collection, employee_collection, get_db
from app.auth import require_auth, require_company_access
from app.utils import get_current_utc

evacuation_bp = Blueprint('evacuation', __name__)


def convert_objectids(obj):
    """Recursively convert ObjectIds to strings"""
    if isinstance(obj, dict):
        return {k: convert_objectids(v) for k, v in obj.items()}
    elif isinstance(obj, list):
        return [convert_objectids(i) for i in obj]
    elif isinstance(obj, ObjectId):
        return str(obj)
    elif isinstance(obj, datetime):
        return obj.isoformat()
    return obj


@evacuation_bp.route('/evacuation-list', methods=['GET'])
@require_company_access
def get_evacuation_list():
    """
    Get real-time list of all currently checked-in visitors.
    
    Critical for emergency response - provides instant headcount of all
    visitors currently on-site with their host and location information.
    
    Query Parameters:
        companyId (required): Company ObjectId
        locationId (optional): Filter by specific location
        includeEmployees (optional): Include checked-in employees (default: false)
    
    Returns:
        JSON with visitors array containing:
        - visitorId, visitorName, phone
        - hostEmployeeName, hostEmployeePhone
        - locationName, checkInTime
        - status (on_site, evacuated, missing)
    """
    try:
        company_id = request.args.get('companyId') or getattr(request, 'company_id', None)
        location_id = request.args.get('locationId')
        include_employees = request.args.get('includeEmployees', 'false').lower() == 'true'
        
        if not company_id:
            return jsonify({'error': 'Company ID is required'}), 400
        
        # Build query for checked-in visits
        query = {
            'status': 'checked_in'
        }
        
        # Handle both ObjectId and string companyId
        try:
            company_oid = ObjectId(company_id)
            query['$or'] = [{'companyId': company_oid}, {'companyId': company_id}]
        except:
            query['companyId'] = company_id
        
        if location_id:
            try:
                query['locationId'] = ObjectId(location_id)
            except:
                query['locationId'] = location_id
        
        # Fetch all checked-in visits
        visits = list(visit_collection.find(query))
        
        evacuation_list = []
        for visit in visits:
            # Get visitor details
            visitor = visitor_collection.find_one({'_id': visit.get('visitorId')})
            
            # Get host employee details
            host = None
            host_id = visit.get('hostEmployeeId')
            if host_id:
                try:
                    host = employee_collection.find_one({'_id': ObjectId(host_id)})
                except:
                    host = employee_collection.find_one({'employeeId': host_id})
            
            evacuation_entry = {
                'visitId': str(visit['_id']),
                'visitorId': str(visit.get('visitorId', '')),
                'visitorName': visit.get('visitorName') or (visitor.get('visitorName') if visitor else 'Unknown'),
                'visitorPhone': visitor.get('phone', '') if visitor else '',
                'visitorOrganization': visitor.get('organization', '') if visitor else '',
                'hostEmployeeId': str(host_id) if host_id else '',
                'hostEmployeeName': visit.get('hostEmployeeName') or (host.get('employeeName') if host else 'Unknown'),
                'hostEmployeePhone': host.get('phone', '') if host else '',
                'hostEmployeeEmail': host.get('email', '') if host else '',
                'locationId': str(visit.get('locationId', '')),
                'locationName': visit.get('locationName', 'Main Building'),
                'checkInTime': visit.get('actualArrival'),
                'checkInMethod': visit.get('checkInMethod', 'unknown'),
                'purpose': visit.get('purpose', ''),
                'accessZones': visit.get('accessAreas', []),
                'evacuationStatus': visit.get('evacuationStatus', 'on_site'),  # on_site, evacuated, missing
                'musterPoint': visit.get('musterPoint'),
                'musterCheckInTime': visit.get('musterCheckInTime')
            }
            
            evacuation_list.append(evacuation_entry)
        
        # Calculate summary
        total_on_site = len(evacuation_list)
        evacuated_count = sum(1 for e in evacuation_list if e['evacuationStatus'] == 'evacuated')
        missing_count = sum(1 for e in evacuation_list if e['evacuationStatus'] == 'missing')
        
        return jsonify({
            'evacuationList': convert_objectids(evacuation_list),
            'summary': {
                'totalOnSite': total_on_site,
                'evacuatedCount': evacuated_count,
                'missingCount': missing_count,
                'accountedFor': evacuated_count,
                'percentAccountedFor': round((evacuated_count / total_on_site * 100) if total_on_site > 0 else 100, 1)
            },
            'generatedAt': get_current_utc().isoformat(),
            'companyId': company_id
        }), 200
        
    except Exception as e:
        print(f"Error getting evacuation list: {e}")
        import traceback
        traceback.print_exc()
        return jsonify({'error': str(e)}), 500


@evacuation_bp.route('/trigger', methods=['POST'])
@require_company_access
def trigger_evacuation():
    """
    Trigger evacuation mode for a company/location.
    
    This will:
    1. Set all current visits to evacuation mode
    2. Send push notifications to all hosts
    3. Display evacuation instructions on kiosks
    4. Generate initial evacuation list
    
    Request Body:
        companyId (required): Company ObjectId
        locationId (optional): Specific location, or all if not specified
        reason (optional): Evacuation reason (fire, drill, security, other)
        musterPoints (optional): List of assembly point locations
    """
    try:
        data = request.json or {}
        company_id = data.get('companyId') or getattr(request, 'company_id', None)
        location_id = data.get('locationId')
        reason = data.get('reason', 'emergency')
        muster_points = data.get('musterPoints', ['Main Assembly Point'])
        
        if not company_id:
            return jsonify({'error': 'Company ID is required'}), 400
        
        # Get or create evacuation record
        db = get_db()
        evacuations_collection = db['evacuations']
        
        # Check if evacuation already in progress
        active_evacuation = evacuations_collection.find_one({
            'companyId': company_id,
            'status': 'in_progress'
        })
        
        if active_evacuation:
            return jsonify({
                'error': 'Evacuation already in progress',
                'evacuationId': str(active_evacuation['_id']),
                'startedAt': active_evacuation.get('startedAt')
            }), 409
        
        # Create evacuation record
        evacuation_doc = {
            '_id': ObjectId(),
            'companyId': company_id,
            'locationId': location_id,
            'reason': reason,
            'musterPoints': muster_points,
            'status': 'in_progress',
            'startedAt': get_current_utc(),
            'startedBy': getattr(request, 'user_id', 'system'),
            'endedAt': None,
            'totalVisitors': 0,
            'totalEvacuated': 0
        }
        
        evacuations_collection.insert_one(evacuation_doc)
        
        # Build query for visits to update
        visit_query = {'status': 'checked_in'}
        try:
            company_oid = ObjectId(company_id)
            visit_query['$or'] = [{'companyId': company_oid}, {'companyId': company_id}]
        except:
            visit_query['companyId'] = company_id
            
        if location_id:
            visit_query['locationId'] = location_id
        
        # Count visitors
        visitor_count = visit_collection.count_documents(visit_query)
        
        # Update all checked-in visits with evacuation status
        visit_collection.update_many(
            visit_query,
            {
                '$set': {
                    'evacuationStatus': 'on_site',  # Will change to 'evacuated' when they check in at muster
                    'evacuationId': evacuation_doc['_id'],
                    'evacuationStartedAt': get_current_utc()
                }
            }
        )
        
        # Update evacuation record with visitor count
        evacuations_collection.update_one(
            {'_id': evacuation_doc['_id']},
            {'$set': {'totalVisitors': visitor_count}}
        )
        
        # TODO: Send push notifications to all hosts
        # TODO: Trigger kiosk evacuation mode
        # TODO: Publish event to platform
        
        return jsonify({
            'message': 'Evacuation triggered successfully',
            'evacuationId': str(evacuation_doc['_id']),
            'totalVisitors': visitor_count,
            'musterPoints': muster_points,
            'status': 'in_progress',
            'startedAt': evacuation_doc['startedAt'].isoformat()
        }), 201
        
    except Exception as e:
        print(f"Error triggering evacuation: {e}")
        import traceback
        traceback.print_exc()
        return jsonify({'error': str(e)}), 500


@evacuation_bp.route('/muster-checkin', methods=['POST'])
@require_company_access
def muster_checkin():
    """
    Check in a visitor at a muster/assembly point.
    
    Called when a visitor arrives at an assembly point during evacuation.
    Updates their evacuation status from 'on_site' to 'evacuated'.
    
    Request Body:
        visitId (required): Visit ObjectId
        musterPoint (required): Which assembly point they checked in at
        method (optional): How they were identified (face, badge, manual)
    """
    try:
        data = request.json or {}
        visit_id = data.get('visitId')
        muster_point = data.get('musterPoint', 'Main Assembly Point')
        method = data.get('method', 'manual')
        
        if not visit_id:
            return jsonify({'error': 'Visit ID is required'}), 400
        
        # Find and update the visit
        visit = visit_collection.find_one({'_id': ObjectId(visit_id)})
        if not visit:
            return jsonify({'error': 'Visit not found'}), 404
        
        if visit.get('status') != 'checked_in':
            return jsonify({'error': 'Visitor is not currently checked in'}), 400
        
        if visit.get('evacuationStatus') == 'evacuated':
            return jsonify({
                'message': 'Visitor already checked in at muster point',
                'musterPoint': visit.get('musterPoint'),
                'musterCheckInTime': visit.get('musterCheckInTime')
            }), 200
        
        # Update visit with muster check-in
        visit_collection.update_one(
            {'_id': ObjectId(visit_id)},
            {
                '$set': {
                    'evacuationStatus': 'evacuated',
                    'musterPoint': muster_point,
                    'musterCheckInTime': get_current_utc(),
                    'musterCheckInMethod': method
                }
            }
        )
        
        # Update evacuation record counter
        if visit.get('evacuationId'):
            db = get_db()
            db['evacuations'].update_one(
                {'_id': visit['evacuationId']},
                {'$inc': {'totalEvacuated': 1}}
            )
        
        # Get visitor name for response
        visitor_name = visit.get('visitorName', 'Unknown')
        
        return jsonify({
            'message': 'Muster check-in successful',
            'visitId': visit_id,
            'visitorName': visitor_name,
            'musterPoint': muster_point,
            'checkedInAt': get_current_utc().isoformat(),
            'method': method
        }), 200
        
    except Exception as e:
        print(f"Error during muster check-in: {e}")
        import traceback
        traceback.print_exc()
        return jsonify({'error': str(e)}), 500


@evacuation_bp.route('/status', methods=['GET'])
@require_company_access
def get_evacuation_status():
    """
    Get current evacuation status and headcount.
    
    Returns real-time status of ongoing evacuation including:
    - Total visitors on-site at start
    - Number evacuated (checked in at muster)
    - Number still missing
    - List of missing visitors
    
    Query Parameters:
        companyId (required): Company ObjectId
        evacuationId (optional): Specific evacuation, or latest if not specified
    """
    try:
        company_id = request.args.get('companyId') or getattr(request, 'company_id', None)
        evacuation_id = request.args.get('evacuationId')
        
        if not company_id:
            return jsonify({'error': 'Company ID is required'}), 400
        
        db = get_db()
        evacuations_collection = db['evacuations']
        
        # Find evacuation
        if evacuation_id:
            evacuation = evacuations_collection.find_one({'_id': ObjectId(evacuation_id)})
        else:
            # Get latest/active evacuation
            evacuation = evacuations_collection.find_one(
                {'companyId': company_id, 'status': 'in_progress'},
                sort=[('startedAt', -1)]
            )
        
        if not evacuation:
            return jsonify({
                'active': False,
                'message': 'No active evacuation',
                'lastEvacuation': None
            }), 200
        
        # Get real-time counts from visits
        evacuated_count = visit_collection.count_documents({
            'evacuationId': evacuation['_id'],
            'evacuationStatus': 'evacuated'
        })
        
        on_site_count = visit_collection.count_documents({
            'evacuationId': evacuation['_id'],
            'evacuationStatus': 'on_site'
        })
        
        # Get list of missing visitors (still on_site status)
        missing_visitors = list(visit_collection.find({
            'evacuationId': evacuation['_id'],
            'evacuationStatus': 'on_site'
        }))
        
        missing_list = []
        for visit in missing_visitors:
            visitor = visitor_collection.find_one({'_id': visit.get('visitorId')})
            host_id = visit.get('hostEmployeeId')
            host = None
            if host_id:
                try:
                    host = employee_collection.find_one({'_id': ObjectId(host_id)})
                except:
                    host = employee_collection.find_one({'employeeId': host_id})
            
            missing_list.append({
                'visitId': str(visit['_id']),
                'visitorName': visit.get('visitorName') or (visitor.get('visitorName') if visitor else 'Unknown'),
                'visitorPhone': visitor.get('phone', '') if visitor else '',
                'hostEmployeeName': host.get('employeeName', '') if host else '',
                'hostEmployeePhone': host.get('phone', '') if host else '',
                'lastKnownLocation': visit.get('locationName', 'Unknown'),
                'checkInTime': visit.get('actualArrival')
            })
        
        total_visitors = evacuated_count + on_site_count
        
        return jsonify({
            'active': evacuation['status'] == 'in_progress',
            'evacuationId': str(evacuation['_id']),
            'reason': evacuation.get('reason', 'emergency'),
            'startedAt': evacuation.get('startedAt'),
            'musterPoints': evacuation.get('musterPoints', []),
            'counts': {
                'total': total_visitors,
                'evacuated': evacuated_count,
                'missing': on_site_count,
                'percentSafe': round((evacuated_count / total_visitors * 100) if total_visitors > 0 else 100, 1)
            },
            'missingVisitors': convert_objectids(missing_list),
            'status': evacuation['status']
        }), 200
        
    except Exception as e:
        print(f"Error getting evacuation status: {e}")
        import traceback
        traceback.print_exc()
        return jsonify({'error': str(e)}), 500


@evacuation_bp.route('/end', methods=['POST'])
@require_company_access
def end_evacuation():
    """
    End an active evacuation.
    
    Marks the evacuation as complete and clears evacuation status from visits.
    Should be called after all-clear is given.
    
    Request Body:
        companyId (required): Company ObjectId
        evacuationId (optional): Specific evacuation ID, or ends current active one
        notes (optional): End notes/summary
    """
    try:
        data = request.json or {}
        company_id = data.get('companyId') or getattr(request, 'company_id', None)
        evacuation_id = data.get('evacuationId')
        notes = data.get('notes', '')
        
        if not company_id:
            return jsonify({'error': 'Company ID is required'}), 400
        
        db = get_db()
        evacuations_collection = db['evacuations']
        
        # Find evacuation
        if evacuation_id:
            evacuation = evacuations_collection.find_one({'_id': ObjectId(evacuation_id)})
        else:
            evacuation = evacuations_collection.find_one({
                'companyId': company_id,
                'status': 'in_progress'
            })
        
        if not evacuation:
            return jsonify({'error': 'No active evacuation found'}), 404
        
        # Get final counts
        final_evacuated = visit_collection.count_documents({
            'evacuationId': evacuation['_id'],
            'evacuationStatus': 'evacuated'
        })
        
        final_missing = visit_collection.count_documents({
            'evacuationId': evacuation['_id'],
            'evacuationStatus': 'on_site'
        })
        
        # End evacuation
        evacuations_collection.update_one(
            {'_id': evacuation['_id']},
            {
                '$set': {
                    'status': 'completed',
                    'endedAt': get_current_utc(),
                    'endedBy': getattr(request, 'user_id', 'system'),
                    'endNotes': notes,
                    'finalEvacuated': final_evacuated,
                    'finalMissing': final_missing
                }
            }
        )
        
        # Clear evacuation status from visits (but keep record of what happened)
        visit_collection.update_many(
            {'evacuationId': evacuation['_id']},
            {
                '$set': {
                    'evacuationEnded': True,
                    'evacuationEndedAt': get_current_utc()
                },
                '$unset': {
                    'evacuationStatus': ''
                }
            }
        )
        
        duration_seconds = (get_current_utc() - evacuation['startedAt']).total_seconds()
        
        return jsonify({
            'message': 'Evacuation ended successfully',
            'evacuationId': str(evacuation['_id']),
            'duration': {
                'seconds': int(duration_seconds),
                'minutes': round(duration_seconds / 60, 1)
            },
            'finalCounts': {
                'evacuated': final_evacuated,
                'missing': final_missing,
                'total': final_evacuated + final_missing
            },
            'endedAt': get_current_utc().isoformat()
        }), 200
        
    except Exception as e:
        print(f"Error ending evacuation: {e}")
        import traceback
        traceback.print_exc()
        return jsonify({'error': str(e)}), 500


@evacuation_bp.route('/history', methods=['GET'])
@require_company_access  
def get_evacuation_history():
    """
    Get history of past evacuations for compliance reporting.
    
    Query Parameters:
        companyId (required): Company ObjectId
        limit (optional): Number of records (default: 20)
        offset (optional): Pagination offset
    """
    try:
        company_id = request.args.get('companyId') or getattr(request, 'company_id', None)
        limit = int(request.args.get('limit', 20))
        offset = int(request.args.get('offset', 0))
        
        if not company_id:
            return jsonify({'error': 'Company ID is required'}), 400
        
        db = get_db()
        evacuations_collection = db['evacuations']
        
        evacuations = list(evacuations_collection.find(
            {'companyId': company_id}
        ).sort('startedAt', -1).skip(offset).limit(limit))
        
        total = evacuations_collection.count_documents({'companyId': company_id})
        
        return jsonify({
            'evacuations': convert_objectids(evacuations),
            'total': total,
            'limit': limit,
            'offset': offset
        }), 200
        
    except Exception as e:
        print(f"Error getting evacuation history: {e}")
        import traceback
        traceback.print_exc()
        return jsonify({'error': str(e)}), 500

File: server/routes/evacuation.js
/**
 * Emergency Evacuation API
 * 
 * Provides endpoints for emergency response and evacuation management:
 * - Real-time list of all on-site visitors
 * - Evacuation mode trigger
 * - Muster point check-in
 * - Headcount and status tracking
 */
const express = require('express');
const router = express.Router();
const { ObjectId } = require('mongodb');
const { collections, getDb } = require('../db');
const { requireCompanyAccess } = require('../middleware/auth');

/**
 * Convert ObjectIds and Dates to strings recursively
 */
function convertObjectIds(obj) {
    if (obj === null || obj === undefined) return obj;
    if (obj instanceof ObjectId) return obj.toString();
    if (obj instanceof Date) return obj.toISOString();
    if (Array.isArray(obj)) return obj.map(convertObjectIds);
    if (typeof obj === 'object') {
        const result = {};
        for (const [key, value] of Object.entries(obj)) {
            result[key] = convertObjectIds(value);
        }
        return result;
    }
    return obj;
}

/**
 * GET /evacuation-list
 * Get real-time list of all currently checked-in visitors
 */
router.get('/evacuation-list', requireCompanyAccess, async (req, res, next) => {
    try {
        const companyId = req.query.companyId || req.companyId;
        const locationId = req.query.locationId;
        const includeEmployees = req.query.includeEmployees === 'true';

        if (!companyId) {
            return res.status(400).json({ error: 'Company ID is required' });
        }

        // Build query for checked-in visits
        const query = { status: 'checked_in' };

        try {
            const companyOid = new ObjectId(companyId);
            query.$or = [{ companyId: companyOid }, { companyId: companyId }];
        } catch {
            query.companyId = companyId;
        }

        if (locationId) {
            try {
                query.locationId = new ObjectId(locationId);
            } catch {
                query.locationId = locationId;
            }
        }

        // Fetch all checked-in visits
        const db = getDb();
        const visits = await db.collection('visits').find(query).toArray();

        const evacuationList = [];
        for (const visit of visits) {
            // Get visitor details
            let visitor = null;
            if (visit.visitorId) {
                try {
                    visitor = await db.collection('visitors').findOne({ _id: new ObjectId(visit.visitorId) });
                } catch {
                    visitor = await db.collection('visitors').findOne({ _id: visit.visitorId });
                }
            }

            // Get host employee details
            let host = null;
            const hostId = visit.hostEmployeeId;
            if (hostId) {
                try {
                    host = await db.collection('employees').findOne({ _id: new ObjectId(hostId) });
                } catch {
                    host = await db.collection('employees').findOne({ employeeId: hostId });
                }
            }

            evacuationList.push({
                visitId: visit._id.toString(),
                visitorId: visit.visitorId?.toString() || '',
                visitorName: visit.visitorName || (visitor?.visitorName || 'Unknown'),
                visitorPhone: visitor?.phone || '',
                visitorOrganization: visitor?.organization || '',
                hostEmployeeId: hostId?.toString() || '',
                hostEmployeeName: visit.hostEmployeeName || (host?.employeeName || 'Unknown'),
                hostEmployeePhone: host?.phone || '',
                hostEmployeeEmail: host?.email || '',
                locationId: visit.locationId?.toString() || '',
                locationName: visit.locationName || 'Main Building',
                checkInTime: visit.actualArrival,
                checkInMethod: visit.checkInMethod || 'unknown',
                purpose: visit.purpose || '',
                accessZones: visit.accessAreas || [],
                evacuationStatus: visit.evacuationStatus || 'on_site',
                musterPoint: visit.musterPoint,
                musterCheckInTime: visit.musterCheckInTime
            });
        }

        // Calculate summary
        const totalOnSite = evacuationList.length;
        const evacuatedCount = evacuationList.filter(e => e.evacuationStatus === 'evacuated').length;
        const missingCount = evacuationList.filter(e => e.evacuationStatus === 'missing').length;

        res.json({
            evacuationList: convertObjectIds(evacuationList),
            summary: {
                totalOnSite,
                evacuatedCount,
                missingCount,
                accountedFor: evacuatedCount,
                percentAccountedFor: totalOnSite > 0 ? Math.round((evacuatedCount / totalOnSite * 100) * 10) / 10 : 100
            },
            generatedAt: new Date().toISOString(),
            companyId
        });

    } catch (error) {
        console.error('Error getting evacuation list:', error);
        next(error);
    }
});

/**
 * POST /trigger
 * Trigger evacuation mode for a company/location
 */
router.post('/trigger', requireCompanyAccess, async (req, res, next) => {
    try {
        const data = req.body || {};
        const companyId = data.companyId || req.companyId;
        const locationId = data.locationId;
        const reason = data.reason || 'emergency';
        const musterPoints = data.musterPoints || ['Main Assembly Point'];

        if (!companyId) {
            return res.status(400).json({ error: 'Company ID is required' });
        }

        const db = getDb();
        const evacuationsCollection = db.collection('evacuations');

        // Check if evacuation already in progress
        const activeEvacuation = await evacuationsCollection.findOne({
            companyId,
            status: 'in_progress'
        });

        if (activeEvacuation) {
            return res.status(409).json({
                error: 'Evacuation already in progress',
                evacuationId: activeEvacuation._id.toString(),
                startedAt: activeEvacuation.startedAt
            });
        }

        const now = new Date();

        // Create evacuation record
        const evacuationDoc = {
            _id: new ObjectId(),
            companyId,
            locationId,
            reason,
            musterPoints,
            status: 'in_progress',
            startedAt: now,
            startedBy: req.userId || 'system',
            endedAt: null,
            totalVisitors: 0,
            totalEvacuated: 0
        };

        await evacuationsCollection.insertOne(evacuationDoc);

        // Build query for visits to update
        const visitQuery = { status: 'checked_in' };
        try {
            const companyOid = new ObjectId(companyId);
            visitQuery.$or = [{ companyId: companyOid }, { companyId: companyId }];
        } catch {
            visitQuery.companyId = companyId;
        }

        if (locationId) {
            visitQuery.locationId = locationId;
        }

        // Count visitors
        const visitsCollection = db.collection('visits');
        const visitorCount = await visitsCollection.countDocuments(visitQuery);

        // Update all checked-in visits with evacuation status
        await visitsCollection.updateMany(visitQuery, {
            $set: {
                evacuationStatus: 'on_site',
                evacuationId: evacuationDoc._id,
                evacuationStartedAt: now
            }
        });

        // Update evacuation record with visitor count
        await evacuationsCollection.updateOne(
            { _id: evacuationDoc._id },
            { $set: { totalVisitors: visitorCount } }
        );

        res.status(201).json({
            message: 'Evacuation triggered successfully',
            evacuationId: evacuationDoc._id.toString(),
            totalVisitors: visitorCount,
            musterPoints,
            status: 'in_progress',
            startedAt: evacuationDoc.startedAt.toISOString()
        });

    } catch (error) {
        console.error('Error triggering evacuation:', error);
        next(error);
    }
});

/**
 * POST /muster-checkin
 * Check in a visitor at a muster/assembly point
 */
router.post('/muster-checkin', requireCompanyAccess, async (req, res, next) => {
    try {
        const data = req.body || {};
        const visitId = data.visitId;
        const musterPoint = data.musterPoint || 'Main Assembly Point';
        const method = data.method || 'manual';

        if (!visitId) {
            return res.status(400).json({ error: 'Visit ID is required' });
        }

        // Find the visit
        const db = getDb();
        const visitsCollection = db.collection('visits');
        const visit = await visitsCollection.findOne({ _id: new ObjectId(visitId) });
        if (!visit) {
            return res.status(404).json({ error: 'Visit not found' });
        }

        if (visit.status !== 'checked_in') {
            return res.status(400).json({ error: 'Visitor is not currently checked in' });
        }

        if (visit.evacuationStatus === 'evacuated') {
            return res.json({
                message: 'Visitor already checked in at muster point',
                musterPoint: visit.musterPoint,
                musterCheckInTime: visit.musterCheckInTime
            });
        }

        const now = new Date();

        // Update visit with muster check-in
        await visitsCollection.updateOne(
            { _id: new ObjectId(visitId) },
            {
                $set: {
                    evacuationStatus: 'evacuated',
                    musterPoint,
                    musterCheckInTime: now,
                    musterCheckInMethod: method
                }
            }
        );

        // Update evacuation record counter
        if (visit.evacuationId) {
            const db = getDb();
            await db.collection('evacuations').updateOne(
                { _id: visit.evacuationId },
                { $inc: { totalEvacuated: 1 } }
            );
        }

        res.json({
            message: 'Muster check-in successful',
            visitId,
            visitorName: visit.visitorName || 'Unknown',
            musterPoint,
            checkedInAt: now.toISOString(),
            method
        });

    } catch (error) {
        console.error('Error during muster check-in:', error);
        next(error);
    }
});

/**
 * GET /status
 * Get current evacuation status and headcount
 */
router.get('/status', requireCompanyAccess, async (req, res, next) => {
    try {
        const companyId = req.query.companyId || req.companyId;
        const evacuationId = req.query.evacuationId;

        if (!companyId) {
            return res.status(400).json({ error: 'Company ID is required' });
        }

        const db = getDb();
        const evacuationsCollection = db.collection('evacuations');

        // Find evacuation
        let evacuation;
        if (evacuationId) {
            evacuation = await evacuationsCollection.findOne({ _id: new ObjectId(evacuationId) });
        } else {
            evacuation = await evacuationsCollection.findOne(
                { companyId, status: 'in_progress' },
                { sort: { startedAt: -1 } }
            );
        }

        if (!evacuation) {
            return res.json({
                active: false,
                message: 'No active evacuation',
                lastEvacuation: null
            });
        }

        // Get real-time counts from visits
        const visitsCollection = db.collection('visits');
        const evacuatedCount = await visitsCollection.countDocuments({
            evacuationId: evacuation._id,
            evacuationStatus: 'evacuated'
        });

        const onSiteCount = await visitsCollection.countDocuments({
            evacuationId: evacuation._id,
            evacuationStatus: 'on_site'
        });

        // Get list of missing visitors
        const missingVisitors = await visitsCollection.find({
            evacuationId: evacuation._id,
            evacuationStatus: 'on_site'
        }).toArray();

        const missingList = [];
        for (const visit of missingVisitors) {
            let visitor = null;
            if (visit.visitorId) {
                try {
                    visitor = await db.collection('visitors').findOne({ _id: new ObjectId(visit.visitorId) });
                } catch { /* ignore */ }
            }

            let host = null;
            if (visit.hostEmployeeId) {
                try {
                    host = await db.collection('employees').findOne({ _id: new ObjectId(visit.hostEmployeeId) });
                } catch {
                    host = await db.collection('employees').findOne({ employeeId: visit.hostEmployeeId });
                }
            }

            missingList.push({
                visitId: visit._id.toString(),
                visitorName: visit.visitorName || (visitor?.visitorName || 'Unknown'),
                visitorPhone: visitor?.phone || '',
                hostEmployeeName: host?.employeeName || '',
                hostEmployeePhone: host?.phone || '',
                lastKnownLocation: visit.locationName || 'Unknown',
                checkInTime: visit.actualArrival
            });
        }

        const totalVisitors = evacuatedCount + onSiteCount;

        res.json({
            active: evacuation.status === 'in_progress',
            evacuationId: evacuation._id.toString(),
            reason: evacuation.reason || 'emergency',
            startedAt: evacuation.startedAt,
            musterPoints: evacuation.musterPoints || [],
            counts: {
                total: totalVisitors,
                evacuated: evacuatedCount,
                missing: onSiteCount,
                percentSafe: totalVisitors > 0 ? Math.round((evacuatedCount / totalVisitors * 100) * 10) / 10 : 100
            },
            missingVisitors: convertObjectIds(missingList),
            status: evacuation.status
        });

    } catch (error) {
        console.error('Error getting evacuation status:', error);
        next(error);
    }
});

/**
 * POST /end
 * End an active evacuation
 */
router.post('/end', requireCompanyAccess, async (req, res, next) => {
    try {
        const data = req.body || {};
        const companyId = data.companyId || req.companyId;
        const evacuationId = data.evacuationId;
        const notes = data.notes || '';

        if (!companyId) {
            return res.status(400).json({ error: 'Company ID is required' });
        }

        const db = getDb();
        const evacuationsCollection = db.collection('evacuations');

        // Find evacuation
        let evacuation;
        if (evacuationId) {
            evacuation = await evacuationsCollection.findOne({ _id: new ObjectId(evacuationId) });
        } else {
            evacuation = await evacuationsCollection.findOne({
                companyId,
                status: 'in_progress'
            });
        }

        if (!evacuation) {
            return res.status(404).json({ error: 'No active evacuation found' });
        }

        const now = new Date();

        // Get final counts
        const visitsCollection = db.collection('visits');
        const finalEvacuated = await visitsCollection.countDocuments({
            evacuationId: evacuation._id,
            evacuationStatus: 'evacuated'
        });

        const finalMissing = await visitsCollection.countDocuments({
            evacuationId: evacuation._id,
            evacuationStatus: 'on_site'
        });

        // End evacuation
        await evacuationsCollection.updateOne(
            { _id: evacuation._id },
            {
                $set: {
                    status: 'completed',
                    endedAt: now,
                    endedBy: req.userId || 'system',
                    endNotes: notes,
                    finalEvacuated,
                    finalMissing
                }
            }
        );

        // Clear evacuation status from visits
        await visitsCollection.updateMany(
            { evacuationId: evacuation._id },
            {
                $set: {
                    evacuationEnded: true,
                    evacuationEndedAt: now
                },
                $unset: {
                    evacuationStatus: ''
                }
            }
        );

        const durationSeconds = (now - evacuation.startedAt) / 1000;

        res.json({
            message: 'Evacuation ended successfully',
            evacuationId: evacuation._id.toString(),
            duration: {
                seconds: Math.round(durationSeconds),
                minutes: Math.round(durationSeconds / 60 * 10) / 10
            },
            finalCounts: {
                evacuated: finalEvacuated,
                missing: finalMissing,
                total: finalEvacuated + finalMissing
            },
            endedAt: now.toISOString()
        });

    } catch (error) {
        console.error('Error ending evacuation:', error);
        next(error);
    }
});

/**
 * GET /history
 * Get history of past evacuations
 */
router.get('/history', requireCompanyAccess, async (req, res, next) => {
    try {
        const companyId = req.query.companyId || req.companyId;
        const limit = parseInt(req.query.limit || '20');
        const offset = parseInt(req.query.offset || '0');

        if (!companyId) {
            return res.status(400).json({ error: 'Company ID is required' });
        }

        const db = getDb();
        const evacuationsCollection = db.collection('evacuations');

        const evacuations = await evacuationsCollection.find({ companyId })
            .sort({ startedAt: -1 })
            .skip(offset)
            .limit(limit)
            .toArray();

        const total = await evacuationsCollection.countDocuments({ companyId });

        res.json({
            evacuations: convertObjectIds(evacuations),
            total,
            limit,
            offset
        });

    } catch (error) {
        console.error('Error getting evacuation history:', error);
        next(error);
    }
});

module.exports = router;
